pkg/engine/compat.go

```diff
@@ -45,107 +45,158 @@ type streamsResultBuilder struct {
 }
 
 func (b *streamsResultBuilder) CollectRecord(rec arrow.Record) {
-	for row := range int(rec.NumRows()) {
-		stream, entry := b.collectRow(rec, row)
-
-		// Ignore rows that don't have stream labels, log line, or timestamp
-		if stream.IsEmpty() || entry.Line == "" || entry.Timestamp.Equal(time.Time{}) {
-			continue
-		}
-
-		// Add the entry to the result builder
-		key := stream.String()
-		idx, ok := b.streams[key]
-		if !ok {
-			idx = len(b.data)
-			b.streams[key] = idx
-			b.data = append(b.data, push.Stream{Labels: key})
-		}
-		b.data[idx].Entries = append(b.data[idx].Entries, entry)
-		b.count++
+	numRows := int(rec.NumRows())
+	if numRows == 0 {
+		return
 	}
-}
 
-func (b *streamsResultBuilder) collectRow(rec arrow.Record, i int) (labels.Labels, logproto.Entry) {
-	var entry logproto.Entry
-	lbs := labels.NewBuilder(labels.EmptyLabels())
-	metadata := labels.NewBuilder(labels.EmptyLabels())
-	parsed := labels.NewBuilder(labels.EmptyLabels())
+	// let's say we have following log entries in rec:
+	// - {labelenv="prod-1", metadatatrace="123-1", parsed="v1"} ts1 line 1
+	// - {labelenv="prod-2", metadatatrace="123-2", parsed="v2"} ts2 line 2
+	// - {labelenv="prod-3", metadatatrace="123-3", parsed="v3"} ts3 line 3
+	// we pre-initialize slices to store column values for all the rows, e.g.:
+	// rows          |    1    |    2    |    3    | ...
+	// ==============+=========+=========+=========+====
+	// timestamps    | r1 ts   | r2 ts   | r3 ts   | ...
+	// lines         | r1 line | r2 line | r3 line | ...
+	// ...
+	// We iterate over the columns and convert the values to our format column by column, e.g.,
+	// first all the timestamps, then all the log lines, etc.
+	// After all the values are collected and converted we transform the columnar representation to a row-based one.
+
+	timestamps := make([]time.Time, numRows)
+	lines := make([]string, numRows)
+	lbsBuilders := make([]*labels.Builder, numRows)
+	metadataBuilders := make([]*labels.Builder, numRows)
+	parsedBuilders := make([]*labels.Builder, numRows)
```

Comment:
Can we re-use these slices so they don't need to be allocated on every CollectRecord() call?
You can call builder.Reset() to re-use the allocated builders in the slice as well.

Reply:
I like the idea. Implemented in the latest commit and updated the benchmark results.

One question though - with this approach the memory allocated for the buffers will not be released for longer periods of time. Do you think it could become a memory issue? I'm worried about the case when:
- there are multiple query executions running at the same time
- all of them wait for the next `pipeline.Read` after at least one `pipeline.Read` was performed and the some records were collected (maybe big records)

As a result all of the result builders for these pipeline will require memory for the buffers, which might lead to an OOM.